File: src/api/categories.ts
import { invoke } from "@tauri-apps/api/core";
import type { Category, SaveCategoryInput } from "../types/category";

export function listCategories(): Promise<Category[]> {
  return invoke<Category[]>("list_categories");
}

export function createCategory(input: SaveCategoryInput): Promise<Category> {
  return invoke<Category>("create_category", { input });
}

export function updateCategory(
  id: number,
  input: SaveCategoryInput,
): Promise<Category> {
  return invoke<Category>("update_category", { id, input });
}

File: src-tauri/src/commands/categories.rs
use tauri::State;

use crate::{
    db::Database,
    error::AppError,
    models::category::{Category, SaveCategoryInput},
    services::categories,
};

use super::{with_authenticated_connection, AuthSession};

#[tauri::command]
pub fn list_categories(
    database: State<'_, Database>,
    session: State<'_, AuthSession>,
) -> Result<Vec<Category>, AppError> {
    with_authenticated_connection(&database, &session, |connection, _| {
        categories::list(connection)
    })
}

#[tauri::command]
pub fn create_category(
    database: State<'_, Database>,
    session: State<'_, AuthSession>,
    input: SaveCategoryInput,
) -> Result<Category, AppError> {
    with_authenticated_connection(&database, &session, |connection, _| {
        categories::create(connection, input)
    })
}

#[tauri::command]
pub fn update_category(
    database: State<'_, Database>,
    session: State<'_, AuthSession>,
    id: i64,
    input: SaveCategoryInput,
) -> Result<Category, AppError> {
    with_authenticated_connection(&database, &session, |connection, _| {
        categories::update(connection, id, input)
    })
}
